### ISSUE TITLE
More precise `__call__` signature is not detected if parent class has `__call__: Callable[..., Any]`

### ISSUE THREAD

**Issue by QuentinSoubeyranAqemia** [microsoft/pyright#11052]
**Describe the bug**
If a parent class contains the annotation `__call__: Callable[..., Any]`, then if a child class assigns to `__call__` a function with a more precise signature (as opposed to defining it with a `def` statement), pyright doesn't pick up the more precise signature.

[pyright playground](https://pyright-play.net/?code=JYWwDg9gTgLgBDAnmYA7A5gKAGZQiBZNdOUSWOAYQEMAbW6gI1oFMAaOAQVUU0wGMGAZyFwACtSgtUMAFx84iuAGIq%2BENPgQArvBgQ4AExYwW-PQAsWcFlDxRMSuAH1n-OrVdxZVD01YA2gB0IRzciAC6jkpg1CLRigLCopQWwLSGABQSUjIAlPIJRizYcNjQAO6SWUIstNgckug%2BaDAcqNQaPkIwUHlwALQAfKQy8k5OUjDaUKhwTQpKwKVIKBhBACoAmmIAos6UABK7lADSAJIAcgDi4xOKru70XgC8ZZXVRUXKxqWPHq5MrV6o0oM1Rm04B0unAen1BiNWndFKolFMZnMFnwDG9UuksnlMBAguUoFUoFkAIwcABEACYaYSIJlqXA6YSpAA3Fh0ZyrFiZYn-Z7OPJAA):
```python
import typing
from typing import Callable, Any

class Parent:

    # Comment out to detect the error
    __call__ : Callable[..., Any]
    pass
    
class Child(Parent):

    def forward(self, arg: int, name: str) -> int:
        return arg

    if typing.TYPE_CHECKING:
        __call__ = forward
    
    #def __call__(self, arg: int, name: str) -> int:
    #    return arg


o = Child()
o.forward(1, "2")
o(1, 2) # should be flagged
reveal_type(o.__call__) # Type of "o.__call__" is "(...) -> Any"
```

If either:
- ` __call__ : Callable[..., Any]` is commented out in the parent
- or the `__call__` override is provided with a `def` instead of an assignment

then PyRight reveals the detailed signature, and reports an error on `o(1,2)`.

Interestingly, [MyPy does pick up the more precise signature](https://mypy-play.net/?mypy=latest&python=3.12&gist=ddb298c0f1f326da30aab96ae881fdc0) in the above example.

This came up when trying to tell VSCode/Pylance that PyTorch's `torch.nn.Module` can be called with the same arguments as their `forward` implementation, using
```python
if typing.TYPE_CHECKING:
    __call__ = forward
```

I found that the `__call__: Callable[..., Any]` annotation in pytorch's source code is interfering with that.

I'm not *quite* sure this is a bug (type-system are hard), although the difference in behavior between assignment and `def` is surprising, and MyPy's behavior is probably more useful in that case.

Cheers!

**Comment by erictraut** [comment]
The behavior you're seeing here is by design. Pyright honors type declarations if they are present. A simple assignment is not a type declaration, so `__call__ = forward` does not override the type declaration in the parent class. If you want to override the declaration, you need to add an annotation in the child class (or use a `def` statement, as you noted).

For an explanation of why pyright and mypy differ in this behavior, refer to [this documentation](https://microsoft.github.io/pyright/#/mypy-comparison?id=class-and-instance-variable-inference).

**Comment by QuentinSoubeyranAqemia** [comment]
Thank you for the clear explanations, and the link to the doc, that behavior/design makes sense. I guess there's no over way than a `def` and a ParamSpec to do it properly.
